service/tasks/monitoring.py

```diff
@@ -287,48 +289,58 @@ def distribute_image_membership(account_driver, cloud_machine, provider):
-        # - SGregory
-        if len(shared_projects_from_main) > len(access_list):
-            shared_projects_from_main = []
-        shared_projects_from_main = [p for p in shared_projects_from_main if p not in shared_projects]
-        shared_projects.extend(shared_projects_from_main)
+    # NOTE: This assumes that the 'owner' (atmosphere user) == 'project_name' (Openstack)
+    # Always include the original application owner
+    shared_project_names = [image_owner]
 
+    shared_projects = account_driver.shared_images_for(cloud_machine.id, None)
     # Extend to include based on projects already granted access to the image
-    shared_group_names.extend([p.name for p in shared_projects if p.name not in shared_group_names])
+    shared_project_names.extend([p.name for p in shared_projects if p.name not in shared_project_names])
 
-
-    #Extend to include new names found by application pattern_match
+    has_machine_request = MachineRequest.objects.filter(
+        new_machine__instance_source__identifier=cloud_machine.id,
+        status__name='completed').last()
+    if has_machine_request:
+        access_list = has_machine_request.get_access_list()
+        # NOTE: This assumes that every name in
+        #      accesslist (AtmosphereUser) == project_name(Openstack)
+        shared_project_names.extend(
+                [name for name in access_list
+                 if name not in shared_project_names])
+        request_provider = has_machine_request.new_machine_provider
+        request_identifier = has_machine_request.new_machine.instance_source.identifier
+        if request_provider != db_machine.provider:
+            main_account_driver = get_account_driver(request_provider)
+            # Extend to include based on information in the machine request
+            request_shared_projects = main_account_driver.shared_images_for(request_identifier, None)
+            request_shared_projects = [p.name for p in request_shared_projects if p.name not in shared_project_names]
+            shared_project_names.extend(request_shared_projects)
+
+    shared_project_names.extend([p.name for p in shared_projects if p.name not in shared_project_names])
+
+    # Extend to include new names found by application pattern_match
     parent_app = db_machine.application_version.application
     matching_users = list(parent_app.get_users_from_access_list().values_list('username', flat=True))
-    if len(matching_users) > 128:
-        # THIS IS A HACK: If there are LOTS of projects shared with the parent application, thats likely
-        # an accident. _ESPECIALLY_ if the list in the access list is smaller. so lets just ignore them.
-        # - SGregory
-        matching_users = []
-    shared_group_names.extend([user for user in matching_users if user not in shared_group_names])
-    return shared_group_names
+    shared_project_names.extend([user for user in matching_users if user not in shared_project_names])
+    return shared_project_names
```

Comment:
Name everything separately and return a concatenation:
```
return <matching_names> + <shared_names> + <share_users>
```
Its not ideal to mutate a variable and move along. Its the difference between me being able to look at the signature of the method (i.e. the return stmt) and more or less know what it does, v.s. having to read the entire body.